armhf: npm install sqlite3 runs into an error

Creating a fresh docker container fails on an armv7l plattform. The error occurs due to an well documented issue with sqlite on armhf plattforms. But my programming skills are not good enough to solve the prolbem :(

My Plattform: Raspberry Pi 3, Debain 
`Linux raspberrypi 4.9.41-v7+ #1023 SMP Tue Aug 8 16:00:15 BST 2017 armv7l GNU/Linux`
The docker-compose.yml
```yaml
version: '3'

services:
  ghost:
    container_name: ghost-1
    image: ghost:1
    restart: always
    ports:
      - "127.0.0.1:2368:2368"
    volumes:
      - "ghost-data:/var/lib/ghost/content"

volumes:
  ghost-data:
```
running compose
```console
# docker-compose up
Creating volume "ghost_ghost-data" with default driver
Creating ghost-1 ... 
Creating ghost-1 ... done
Attaching to ghost-1
ghost-1  | Knex: run
ghost-1  | $ npm install sqlite3 --save
ghost-1  | Error: Cannot find module 'sqlite3'
ghost-1  |     at Function.Module._resolveFilename (module.js:469:15)
ghost-1  |     at Function.Module._load (module.js:417:25)
ghost-1  |     at Module.require (module.js:497:17)
ghost-1  |     at require (internal/module.js:20:19)
ghost-1  |     at Client_SQLite3._driver (/usr/local/lib/node_modules/knex-migrator/node_modules/knex/lib/dialects/sqlite3/index.js:75:12)
ghost-1  |     at Client_SQLite3.initializeDriver (/usr/local/lib/node_modules/knex-migrator/node_modules/knex/lib/client.js:219:26)
ghost-1  |     at Client_SQLite3.Client (/usr/local/lib/node_modules/knex-migrator/node_modules/knex/lib/client.js:113:10)
ghost-1  |     at new Client_SQLite3 (/usr/local/lib/node_modules/knex-migrator/node_modules/knex/lib/dialects/sqlite3/index.js:58:20)
ghost-1  |     at Knex (/usr/local/lib/node_modules/knex-migrator/node_modules/knex/lib/index.js:60:34)
ghost-1  |     at Object.connect (/usr/local/lib/node_modules/knex-migrator/lib/database.js:25:12)
ghost-1  |     at KnexMigrator.init (/usr/local/lib/node_modules/knex-migrator/lib/index.js:71:32)
ghost-1  |     at KnexMigrator.migrate (/usr/local/lib/node_modules/knex-migrator/lib/index.js:161:21)
ghost-1  |     at /usr/local/lib/node_modules/knex-migrator/bin/knex-migrator-migrate:26:29
ghost-1  |     at tryCatcher (/usr/local/lib/node_modules/knex-migrator/node_modules/bluebird/js/release/util.js:16:23)
ghost-1  |     at Promise._settlePromiseFromHandler (/usr/local/lib/node_modules/knex-migrator/node_modules/bluebird/js/release/promise.js:512:31)
ghost-1  |     at Promise._settlePromise (/usr/local/lib/node_modules/knex-migrator/node_modules/bluebird/js/release/promise.js:569:18)
ghost-1  |     at Promise._settlePromise0 (/usr/local/lib/node_modules/knex-migrator/node_modules/bluebird/js/release/promise.js:614:10)
ghost-1  |     at Promise._settlePromises (/usr/local/lib/node_modules/knex-migrator/node_modules/bluebird/js/release/promise.js:693:18)
ghost-1  |     at Promise._fulfill (/usr/local/lib/node_modules/knex-migrator/node_modules/bluebird/js/release/promise.js:638:18)
ghost-1  |     at Promise._resolveCallback (/usr/local/lib/node_modules/knex-migrator/node_modules/bluebird/js/release/promise.js:432:57)
ghost-1  |     at Promise._settlePromiseFromHandler (/usr/local/lib/node_modules/knex-migrator/node_modules/bluebird/js/release/promise.js:524:17)
ghost-1  |     at Promise._settlePromise (/usr/local/lib/node_modules/knex-migrator/node_modules/bluebird/js/release/promise.js:569:18)
ghost-1 exited with code 1
```

--- a/1/alpine/Dockerfile
+++ b/1/alpine/Dockerfile
@@ -9,15 +9,15 @@ RUN apk add --no-cache \
 # add "bash" for "[["
 		bash
 
-ENV NPM_CONFIG_LOGLEVEL warn
 ENV NODE_ENV production
-ENV GHOST_CLI_VERSION 1.2.0
-ENV GHOST_VERSION 1.16.2
 
-RUN npm install -g "ghost-cli@$GHOST_CLI_VERSION" knex-migrator@latest
+ENV GHOST_CLI_VERSION 1.2.0
+RUN npm install -g "ghost-cli@$GHOST_CLI_VERSION"
 
 ENV GHOST_INSTALL /var/lib/ghost
 ENV GHOST_CONTENT /var/lib/ghost/content
+
+ENV GHOST_VERSION 1.16.2
 
 RUN set -ex; \
 	mkdir -p "$GHOST_INSTALL"; \
@@ -37,7 +37,16 @@ RUN set -ex; \
 # need to save initial content for pre-seeding empty volumes
 	mv "$GHOST_CONTENT" "$GHOST_INSTALL/content.orig"; \
 	mkdir -p "$GHOST_CONTENT"; \
-	chown node:node "$GHOST_CONTENT"
+	chown node:node "$GHOST_CONTENT"; \
+	\
+# symlink knex-migrator bins into PATH
+# we want these from the context of Ghost's "node_modules" directory (instead of doing "npm install -g knex-migrator") so they can share the DB driver modules
+	for f in "$GHOST_INSTALL/current/node_modules/.bin/knex-migrator"*; do \
+		[ -x "$f" ]; \
+		ln -svf "$f" /usr/local/bin/; \
+	done
+
+# TODO multiarch sqlite3 (once either "node:6-alpine" has multiarch or we switch to a base that does)
 
 WORKDIR $GHOST_INSTALL
 VOLUME $GHOST_CONTENT

--- a/1/debian/Dockerfile
+++ b/1/debian/Dockerfile
@@ -14,15 +14,15 @@ RUN set -x \
 	&& chmod +x /usr/local/bin/gosu \
 	&& gosu nobody true
 
-ENV NPM_CONFIG_LOGLEVEL warn
 ENV NODE_ENV production
-ENV GHOST_CLI_VERSION 1.2.0
-ENV GHOST_VERSION 1.16.2
 
-RUN npm install -g "ghost-cli@$GHOST_CLI_VERSION" knex-migrator@latest
+ENV GHOST_CLI_VERSION 1.2.0
+RUN npm install -g "ghost-cli@$GHOST_CLI_VERSION"
 
 ENV GHOST_INSTALL /var/lib/ghost
 ENV GHOST_CONTENT /var/lib/ghost/content
+
+ENV GHOST_VERSION 1.16.2
 
 RUN set -ex; \
 	mkdir -p "$GHOST_INSTALL"; \
@@ -42,7 +42,35 @@ RUN set -ex; \
 # need to save initial content for pre-seeding empty volumes
 	mv "$GHOST_CONTENT" "$GHOST_INSTALL/content.orig"; \
 	mkdir -p "$GHOST_CONTENT"; \
-	chown node:node "$GHOST_CONTENT"
+	chown node:node "$GHOST_CONTENT"; \
+	\
+# symlink knex-migrator bins into PATH
+# we want these from the context of Ghost's "node_modules" directory (instead of doing "npm install -g knex-migrator") so they can share the DB driver modules
+	for f in "$GHOST_INSTALL/current/node_modules/.bin/knex-migrator"*; do \
+		[ -x "$f" ]; \
+		ln -svf "$f" /usr/local/bin/; \
+	done
+
+RUN set -eux; \
+# force install "sqlite3" manually since it's an optional dependency of "ghost"
+# (which means that if it fails to install, like on ARM/ppc64le/s390x, the failure will be silently ignored and thus turn into a runtime error instead)
+# see https://github.com/TryGhost/Ghost/pull/7677 for more details
+	cd "$GHOST_INSTALL/current"; \
+# scrape the expected version of sqlite3 directly from Ghost itself
+	sqlite3Version="$(npm view . optionalDependencies.sqlite3)"; \
+	if ! gosu node npm install "sqlite3@$sqlite3Version"; then \
+# must be some non-amd64 architecture pre-built binaries aren't published for, so let's install some build deps and do-it-all-over-again
+		savedAptMark="$(apt-mark showmanual)"; \
+		apt-get update; \
+		apt-get install -y --no-install-recommends python make gcc g++ libc-dev; \
+		rm -rf /var/lib/apt/lists/*; \
+		\
+		gosu node npm install "sqlite3@$sqlite3Version" --build-from-source; \
+		\
+		apt-mark showmanual | xargs apt-mark auto > /dev/null; \
+		[ -z "$savedAptMark" ] || apt-mark manual $savedAptMark; \
+		apt-get purge -y --auto-remove; \
+	fi
 
 WORKDIR $GHOST_INSTALL
 VOLUME $GHOST_CONTENT

